plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,654 @@
+					    "hub opcode 0x%x requires metadata",
+					    opcode);
+				return FALSE;
+			}
+			if (!fu_dell_dock_usb4_hub_set_mmio (device,
+							     MBOX_REG_METADATA,
+							     metadata, error)) {
+				g_prefix_error (error,
+						"failed to write metadata: %s",
+						metadata);
+				return FALSE;
+			}
+			break;
+		default:
+			g_prefix_error (error, "invalid hub opcode: 0x%x",
+					opcode);
+			return FALSE;
+	}
+
+	/* write the operation and poll completion or error */
+	if (!fu_dell_dock_usb4_hub_set_mmio (device, MBOX_REG, buf, error))
+		return FALSE;
+
+	for (i = 0; i < 100; i++) {
+		if (!fu_dell_dock_usb4_hub_get_mmio (device, MBOX_REG, buf,
+						     error))
+			return FALSE;
+
+		/* Error status bit set, operation failed */
+		if (regx->status & MBOX_ERROR) {
+			g_prefix_error (error,
+					"hub failed to execute "
+					"operation 0x%x, "
+					"stat 0x%x, opcode 0x%x\n",
+					opcode,
+					regx->status, regx->opcode);
+			return FALSE; /* FIXME error code */
+		}
+		/* Success, opetaion completed */
+		if (!(regx->status & MBOX_OPVALID)) {
+			*status = regx->status;
+			return TRUE;
+		}
+		nanosleep(&poll_delay, NULL);
+	}
+	g_prefix_error (error, "Timed out");
+	return FALSE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_nvm_read (FuDevice *device, guint8 *buf, guint32 length,
+			        guint32 nvm_addr, GError **error)
+{
+	guint8 tmpbuf[64];
+	guint8 op_status;
+
+	while (length > 0) {
+		guint32	unaligned_bytes = nvm_addr % 4;
+		guint32	padded_len;
+		guint32	nbytes;
+		guint8	metadata[4];
+
+		if (length + unaligned_bytes < 64) {
+			nbytes = length;
+			padded_len = unaligned_bytes + length;
+
+			/* align end to full dword boundary */
+			if (padded_len % 4)
+				padded_len = (padded_len & ~0x3) + 4;
+		} else {
+			padded_len = 64;
+			nbytes = padded_len - unaligned_bytes;
+		}
+
+		/* set nvm read offset in dwords */
+		fu_dell_dock_usb4_write_u32to_buf (NVM_OFFSET_TO_METADATA(nvm_addr),
+						   metadata);
+
+		/* and length field in dwords, note 0 means 16 dwords */
+		metadata[3] = (padded_len / 4) & 0xf;
+
+		/* ask hub to read up to 64 bytes from NVM to mbox data regs */
+		if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_READ,
+						      metadata, &op_status,
+						      error)) {
+			g_prefix_error (error, "hub firmware NVM read error");
+			return FALSE;
+		}
+		/* FIXME take a look at op_status here */
+
+		/* read the data from mbox data regs into our buffer */
+		if (!fu_dell_dock_usb4_mbox_data_read (device, tmpbuf,
+						       padded_len, error)) {
+			g_prefix_error (error,
+					"hub firmware mbox data read error");
+			return FALSE;
+		}
+		memcpy(buf, tmpbuf + unaligned_bytes, nbytes);
+
+		buf += nbytes;
+		nvm_addr += nbytes;
+		length -= nbytes;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_nvm_write (FuDevice *device, const guint8 *buf,
+				 guint32 length, guint32 nvm_addr,
+				 GError **error)
+{
+	guint8 metadata[4];
+	guint8 op_status;
```

Comment:
You need to initialize this to something.  Otherwise this can be a problem reading a garbage variable if it never got overwritten in `fu_dell_dock_usb4_hub_operation`